routing/index.ts

```diff
@@ -0,0 +1,183 @@
+import {createAction} from '../actions';
+import {withTask} from '../tasks';
+import {HISTORY_PUSH_TASK} from '../history';
+
+type ComponentType = any;
+
+type RouteDefinition = {
+  [route: string]: ComponentType | [ComponentType, RouteDefinition]
+}
+
+/*
+ * Example usage:
+ *
+ * createRouter({
+ *   '/': Index,
+ *   '/users': [User, {
+ *     '/'
+ *   }]
+ * })
+ */
+export function createRouter(routes: RouteDefinition, onChange = (state) => state) {
+  const LOCATION_CHANGE = createAction('LOCATION_CHANGE');
+  const HISTORY_PUSH = createAction('HISTORY_PUSH');
+
+  const DEFAULT_CHILD = routes['/'];
+  const DEFUALT_ROUTES = [{component: DEFAULT_CHILD, params: {}}];
+
+  const INITIAL_STATE = {
+    pathname: '',
+    routes: DEFUALT_ROUTES
+  };
+
+  const handlers = {
+    [HISTORY_PUSH as any as string]: (state, pathname) =>
+      withTask(updateRoute(state, pathname), HISTORY_PUSH_TASK(pathname)),
+    [LOCATION_CHANGE as any as string]: (state, {pathname}) => updateRoute(state, pathname)
+  };
+
+  function updateRoute(state, pathname) {
+    const maybeMatch = deepMatchRoutes(routes, pathname);
+    const nextState = Object.assign({}, state, {
+      routing: {pathname, routes: maybeMatch || DEFUALT_ROUTES}
+    });
+    return onChange(nextState);
+  }
+
+  return {handlers, INITIAL_STATE, HISTORY_PUSH, LOCATION_CHANGE};
+}
+
+// returns an array of pairs of {component, params}
+function deepMatchRoutes(routes, pathname) {
+  const routePatterns = Object.keys(routes);
+  let index = 0;
+  for (const pattern of routePatterns) {
```

Comment:
Yes. `for of` lets us early return (see line 65, 68). I also prefer the functional form generally, but I couldn't think of a nicer way to accomplish the same with `array.proto.find` + `map` or `some`.  Feel free to suggest something.